### ISSUE TITLE
Using without package?

### ISSUE THREAD

**Issue by nstrayer** [ramnathv/htmlwidgets#305]
Is it possible to build an HTML widget without needing it to be its own package? 

I am currently using HTML widgets to build custom visualization tools for proprietary medical data and have run into the issue of deploying to RStudio Connect as it needs packrat to be able to find the package source. What I'm planning to do is just convert it to a plain shiny module, but I will miss some of the nice scaffolding features that widgets provide. 

I apologize if this has been answered somewhere already, my searches came up empty. 

Thanks for a wonderful package!

**Comment by timelyportfolio** [comment]
We started down this path a while ago in https://github.com/ramnathv/htmlwidgets/issues/176 but never resolved.  Glad you are reviving the discussion.

**Comment by nstrayer** [comment]
I need to familiarize myself with the internals but could this be accomplished by providing a path to libraries and to the main js file to a constructor which returns a list of the main function and the two shiny functions? I'm assuming the package build step is there to gather and inject the JavaScript to a single source?

**Comment by timelyportfolio** [comment]
@nstrayer I'd like to pursue this further as it relates to @fbreitweiser #304 and could provide a good mechanism for quick inclusion of `htmlwidget` functionality.  From what I remember, one of the primary reasons for requiring a full package was the desire for local `dependencies` in a knowable location.  However, with many packages, such as `d3r` and `rmarkdown` offering dependencies and the difficulty specifying order of dependencies in the `yaml` configuration, I am not sure this is as necessary as we originally thought.

Really, if we remove the dependency requirement, we should be able to achieve this objective (unless I am missing something - likely).  So, hopefully this will push the conversation forward.

There is one place where we go looking for file dependencies
- [createWidget lines](https://github.com/ramnathv/htmlwidgets/blob/master/R/htmlwidgets.R#L209-L212) where we call [`widget_dependencies()` lines] which ultimately depends on (https://github.com/ramnathv/htmlwidgets/blob/master/R/htmlwidgets.R#L237-L239) [`getDependency()` lines](https://github.com/ramnathv/htmlwidgets/blob/master/R/utils.R#L41-L80).

One option would be to look for the files as we currently do, but if they do not exist then proceed without error assuming the user or widget builder is intentionally breaking the convention.

**Comment by jjallaire** [comment]
I think you could create htmlwigets outside of a package except for the requirement of the YAML file. If this was optional rather than required then any dependencies could also be provided via the dependencies argument to `createWidget`

**Comment by timelyportfolio** [comment]
Yep, something like this in place of [lines](https://github.com/ramnathv/htmlwidgets/blob/master/R/utils.R#L48-L54) seems to work to remove the necessity of `yaml`.

```
  # if yaml does not exist then assume no dependencies
  widgetDep <- list()
  if(file.exists(system.file(config, package = package))) {
    config = yaml::yaml.load_file(
      system.file(config, package = package)
    )
    widgetDep <- lapply(config$dependencies, function(l){
      l$src = system.file(l$src, package = package)
      do.call(htmlDependency, l)
    })
  }
```

I'll submit a pull for better testing later today.

**Comment by timelyportfolio** [comment]
Then, the next step toward complete freedom would be to remove the necessity of a `JavaScript` binding.  For this pursuing a similar approach to above, we could replace [lines](https://github.com/ramnathv/htmlwidgets/blob/master/R/utils.R#L56-L79) with 


```
  # If js binding does not exist then assume provided through
  #  some other mechanism.  Note, this is a very special case.
  bindingDep <- NULL
  if(file.exists(system.file(jsfile, package = package))) {
    bindingDir <- system.file("htmlwidgets", package = package)
    argsDep <- NULL
    bindingDep <- do.call(htmlDependency, c(list(
      paste0(name, "-binding"), packageVersion(package),
      bindingDir, script = basename(jsfile)
    ), argsDep))
  }

  c(
    list(htmlDependency("htmlwidgets", packageVersion("htmlwidgets"),
                        src = system.file("www", package="htmlwidgets"),
                        script = "htmlwidgets.js"
    )),
    widgetDep,
    list(bindingDep)
  )
```

**Comment by ramnathv** [comment]
@timelyportfolio This makes sense and should work. Looking forward to your PR.

**Comment by timelyportfolio** [comment]
As a test, I manually removed the `yaml` from my `sunburstR` package in library and prepended `_` to the `sund2b.js`.  Then, I successfully created an `htmlwidget` without a package using

```
library(htmltools)

browsable(
  tagList(
    sund2b(sequences),
    htmlDependency(
      name = "sund2b",
      version = "0.1.0",
      src = c(file="C://Users/kent.russell/Documents"),
      script = "_sund2b.js"
    )
  )
)
```

**Comment by timelyportfolio** [comment]
Even more aggressive, I was able to 

```
library(htmltools)

browsable(
  tagList(
    tags$script("..."), # with body of sund2b.js binding
    sund2b(sequences)
  )
)
```

**Comment by timelyportfolio** [comment]
closed by #306; feel free to reopen for discussion or issues
